pkg/distributor/ha_tracker.go

```diff
@@ -202,10 +202,55 @@ func newHATracker(cfg HATrackerConfig, limits haTrackerLimits, reg prometheus.Re
 		t.client = client
 	}
 
-	t.Service = services.NewBasicService(nil, t.loop, nil)
+	t.Service = services.NewBasicService(t.syncHAStateOnStart, t.loop, nil)
 	return t, nil
 }
 
+// Will update the internal cache to be in sync with the KV Store
+func (h *haTracker) syncHAStateOnStart(ctx context.Context) error {
+	if !h.cfg.EnableHATracker {
+		return nil
+	}
+
+	// haTracker holds HATrackerConfig with the prefix.
+	keys, err := h.client.List(ctx, "")
+	if err != nil {
+		level.Warn(h.logger).Log("msg", "syncHAStateOnStart: failed to list replica keys", "err", err)
+		return err
+	}
+
+	if len(keys) == 0 {
+		level.Warn(h.logger).Log("msg", "syncHAStateOnStart: no keys for HA tracker prefix", "err", err)
+		return nil
+	}
+
+	// TODO what is the benefit to allowing an a concurrent/parallel process here
+	// What is the amount fo keys we are getting back? Is it inefficient to serially allow this behavior?
+	for i := 0; i < len(keys); i++ {
+		if ctx.Err() != nil {
+			return ctx.Err()
```

Comment:
we've had some issues with improper context cancellation and vague context cancellation messages. Let's use `context.Cause(ctx)` instead of `ctx.Err`, and also wrap the error into something saying when the error happened: `fmt.Errorf("syncing HA tracker state on startup: %w")`

Reply:
ahh gotcha, makes sense. I'll make that change, ty.